impls/src/tor/proxy.rs

```diff
@@ -0,0 +1,198 @@
+		let mut host: String;
+		let str_port: Option<String>;
+		let mut split = vec![];
+		let address = TorProxy::remove_whitespace(addr);
+		if address.starts_with('[') {
+			split = address.split_inclusive("]:").collect();
+			host = split[0].into();
+			host.push_str("]");
+			str_port = Some(split[1].into());
+		} else if address.contains(":") {
+			split = address.split(":").collect();
+			host = split[0].into();
+			str_port = Some(split[1].into());
+		} else {
+			host = address.to_string();
+			str_port = None;
+		};
+		if (split.len()) > 2 {
+			let msg = format!("Error parsing the address: {}", addr);
+			return Err(ErrorKind::TorProxy(msg).into());
+		}
+		Ok((host, str_port))
+	}
+
+	pub fn parse_address(addr: &str) -> Result<(String, Option<u16>), Error> {
+		let (host, str_port) = TorProxy::parse_ip_dns(&addr)?;
+		let host = Host::parse(&host)
+			.map_err(|_e| ErrorKind::TorProxy(format!("Invalid host address: {}", host)))?;
+		let port = if let Some(p) = str_port {
+			let res = p.parse::<u16>().map_err(|_e| {
+				ErrorKind::TorProxy(format!("Invalid port in the server address: {}", p))
+			})?;
+			Some(res)
+		} else {
+			None
+		};
+		Ok((host.to_string(), port))
+	}
+
+	pub fn to_hashmap(self) -> Result<HashMap<String, String>, Error> {
+		let mut hm: HashMap<String, String> = HashMap::new();
+		if let Some(ports) = self.allowed_port {
+			let mut allowed_ports = "".to_string();
+			let last_port = ports.last().unwrap().to_owned();
+			for port in ports.clone() {
+				allowed_ports.push_str(format!("*:{}", port).as_str());
+				if port != last_port {
+					allowed_ports.push_str(",");
+				}
+			}
+			hm.insert(
+				"ReachableAddresses".to_string(),
+				format!("{}", allowed_ports.clone()),
+			);
+		}
+
+		let transport = match self.transport {
+			Some(t) => t.to_lowercase(),
+			None => return Ok(hm),
+		};
+		match transport.as_str() {
+			"socks4" => {
+				hm.insert("Socks4Proxy".to_string(), self.address.unwrap());
+				Ok(hm)
+			}
+			"socks5" => {
+				hm.insert("Socks5Proxy".to_string(), self.address.unwrap());
+
+				if let Some(s) = self.username {
+					hm.insert("Socks5ProxyUsername".to_string(), s);
+				}
+				if let Some(s) = self.password {
+					hm.insert("Socks5ProxyPassword".to_string(), s);
+				}
+				Ok(hm)
+			}
+			"http" | "https" | "http(s)" => {
+				hm.insert("HTTPSProxy".to_string(), self.address.unwrap());
+
+				if let Some(user) = self.username {
+					let pass = self.password.unwrap_or("".to_string());
+					hm.insert(
+						"HTTPSProxyAuthenticator".to_string(),
+						format!("{}:{}", user, pass),
+					);
+				}
+				Ok(hm)
+			}
+			_ => Ok(hm),
+		}
+	}
+}
+
+impl TryFrom<TorProxyConfig> for TorProxy {
+	type Error = Error;
+
+	fn try_from(tb: TorProxyConfig) -> Result<Self, Self::Error> {
+		if let Some(t) = tb.transport {
+			let transport = t.to_lowercase();
+			match transport.as_str() {
+				"socks4" | "socks5" | "http" | "https" | "http(s)" => {
```

Comment:
Do we need to include 'socks5h' here? (See #629)

Reply:
The match that you commented, is just in order to parse the user input in the config file for him to choose which type of proxy he want to set. You could refer to the tor browser settings for reference. And write accordingly in the torrc config the transport, cf https://github.com/mimblewimble/grin-wallet/pull/617/files#diff-260d6fd89f30bf49927a33dffd4a77eb74479cb3b919f1f818478d1db8fa1e78R115-R144